Improve signal-to-noise ratio of cms export diff tool output

mpelzsherman opened the pull request:
This PR filters out the array comparisons from the deep-diff, and instead stores the arrays for comparison a more targetted method.

Other optimizations are included to reduce false positives and noise in the tool output, e.g. excluding more fields.

There are still some false positives in the array data, but there are many improvements here. We will continue to iterate.

mpelzsherman committed:
Check arrays in diff tool

mpelzsherman committed:
Omit arrays from deep-diff comparison. Use lodash insteadh.

mpelzsherman committed:
Move deep-diff to devDependencies

U-DON commented on script/cms/compare-json.js:
```diff
@@ -182,24 +225,48 @@ const compareJson = (baseGraphQlObject, baseCmsExportObject) => {
-  return deepDiff(graphQlObject, cmsExportObject, (diffPath, key) => {
+  const diffs = deepDiff(graphQlObject, cmsExportObject, (diffPath, key) => {
+    // Store arrays for comparison by alternate method.
+    const cmsObject = Object.byArray(cmsExportObject, diffPath);
+    const graphQLObject = Object.byArray(cmsExportObject, diffPath);
```
Is the `get` function from `src/platform/utilities/data/get.js` suitable for this? It would return `undefined` (or some other default) for paths that don't exist on the object, but I think in those cases, we should return `false` for the equailty check anyway.

mpelzsherman replied:
🤦  I forgot about `get`! Thanks, I'll use that instead!!

mpelzsherman replied:
I ran into problems using src/platform/utilities/data/get.js because we're not in a module, so I used lodash get instead.

theodur commented on script/cms/compare-json.js:
```diff
@@ -182,24 +225,48 @@ const compareJson = (baseGraphQlObject, baseCmsExportObject) => {
+  const diffs = deepDiff(graphQlObject, cmsExportObject, (diffPath, key) => {
+    // Store arrays for comparison by alternate method.
+    const cmsObject = Object.byArray(cmsExportObject, diffPath);
+    const graphQLObject = Object.byArray(cmsExportObject, diffPath);
```
Shouldn't this be:
```suggestion
    const graphQLObject = Object.byArray(graphQlObject, diffPath);
```
Otherwise `cmsObject` and `graphQLObject` will always be the same value.

mpelzsherman replied:
Arrrrgh! Good catch, thanks!

theodur commented on script/cms/compare-json.js:
```diff
@@ -154,6 +157,49 @@ const getParentNode = (
+  cmsExportArray.forEach((cmsItem, index) => {
+    if (!isEqual(cmsItem, graphQLArray[index])) {
+      console.log('ARRAY DIFF FOUND FOR', dataPath, 'index', index);
+      console.log('CMS item:', JSON.stringify(cmsItem, null, 2));
```
Can we output this to a file maybe? I imagine it will be a lot displayed to the console.

mpelzsherman replied:
That can be done using a UNIX redirect, if need be. So far I am not seeing any output at all. If that turns out to be erroneous I will reconsider. Ultimately we should probably put this data into the same format as the other diffs, but since I wasn't seeing anything I figured a simple console.log would suffice to unblock facilities and PW teams for now.

theodur replied:
To get this data to have the same format as the other results you have the `compareArrays` function return an object similar to what we return, i.e. 
```js
{
    diffType: "Array Diff",
    path: dataPath,
    graphQL: graphQLArray[index]
    cmsExport: cmsItem
}
  ```

 and call the `compareArrays` function directly in the `if (isArray)` statement, and push the results to an array. Then you can append that array to the `diffs` array if it's not `undefined`. This way, the array differences would be outputted in the same diff file.

U-DON commented on src/applications/health-care-questionnaire/questionnaire/api/submit-questionnaire.js:
```diff
@@ -1,4 +1,4 @@
-import { submitToUrl } from 'platform/forms-system/src/js/actions';
+// import { submitToUrl } from 'platform/forms-system/src/js/actions';
```
Are these changes relevant?

mpelzsherman committed:
Fix bug in array comparisons.

Use lodash get to get object by path.

Ignore keysToIgnore in array comparisons.

theodur committed:
Make separate files for array comparisons

mpelzsherman committed:
Add entityBundle and entityId to keysToIgnore

Use lodash transform to recursively remove ignored keys

Truncate array diffs to 100

mpelzsherman committed:
Add JSON.stringify check to remove false positives caused by isEqual

mpelzsherman committed:
Remove "processed" from fieldErrorMessage in paragraph-react_widget.js

mpelzsherman committed:
add breadcrumb to keysToIgnore

mpelzsherman committed:
strip ignored keys recursively (WIP)

mpelzsherman committed:
Remove entityPublished from event-listing transformer

theodur committed:
Add array length diffs to output

theodur committed:
Strip ignored keys from the non array diff objects

kelsonic reviewed (approved):
LGTM 🎉

theodur committed:
Use deep-diff on objects in array comparisons
